eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+
+    # TODO: validate activation_layer dimensions (is it possible to perform Grad-CAM on it?)
+    return activation_layer
+
+
+def search_layer_backwards(estimator, condition):
+    """
+    Search for a layer in ``estimator`` backwards (starting from output layer),
+    checking if the layer is suitable with the callable ``condition``,
+    where condition takes ``estimator`` and ``index`` arguments.
+    
+    Returns
+    -------
+    layer : object
+        A suitable ``keras.layers.Layer`` instance.
+
+    Raises
+    ------
+    ValueError :
+        If suitable layer can not be found.
+    """
+    # we assume that this is a simple feedforward network
+    # linear search in reverse
+    i = len(estimator.layers)-1
+    while -1 < i and not condition(estimator, i):
+        i -= 1
+    if -1 < i:
+        # linear search succeeded
+        return estimator.get_layer(index=i)
+    else:
+        raise ValueError('Could not find a suitable target layer automatically.')
+
+
+def is_suitable_activation_layer(estimator, i):
+    """
+    Check whether
+    the layer at index ``i`` matches what is required 
+    by ``estimator``.
+    
+    Matching is done by
+    * checking the rank of the layer.
+
+    Returns
+    -------
+    is suitable : boolean
+        whether the layer matches what is needed
+    """
+    # TODO: experiment with this, using many models and images, to find what works best
+    # Some ideas: 
+    # check layer type, i.e.: isinstance(l, keras.layers.Conv2D)
+    # check layer name
+    l = estimator.get_layer(index=i)
+    # a check that asks 'can we resize this activation layer over the image?'
+    rank = len(l.output_shape)
+    required_rank = len(estimator.input_shape)
+    return rank == required_rank
+
+
+def get_target_prediction(model, x, targets):
+    """
+    Get a prediction ID from ``targets``.
+
+    See documentation of ``explain_prediction_keras`` for explanation of ``targets``.
+    
+    Returns
+    -------
+    prediction id : int
+    """
+    # TODO: take in a single target as well, not just a list, 
+    # consider changing signature / types for explain_prediction generic function
+    # TODO: need to find a way to show the label for the passed prediction 
+    # as well as its probability
+
+    # TODO: maybe do the sum / loss in this function instead of grad_cam. Return a tensor.
+    # This would be consistent with what is done in https://github.com/ramprs/grad-cam/blob/master/misc/utils.lua
+    # https://github.com/ramprs/grad-cam/blob/master/classification.lua
+    # https://github.com/torch/nn/blob/master/doc/module.md
+    if isinstance(targets, list):
+        # take the first prediction from the list
+        predicted_idx = targets[0]
+        # TODO: use all predictions in the list
+        # TODO: validate list contents
+    elif targets is None:
+        predictions = model.predict(x)
```

Comment:
Would be great to avoid doing prediction twice (here and when we're explaining the output) if possible, as it's a costly operation

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/103d9f03a09e7e8328e78f9c1fe8ff2935db0f42